File: LLM/paged_attention.py
"""
PagedAttention — vLLM 风格离散物理块 KV 缓存与页表寻址 Demo。

核心思想
--------
KV Cache 按固定大小 Block 切分，逻辑连续、物理离散。
BlockManager 维护全局物理显存池；每个请求持有一张 block_table（逻辑块 -> 物理块 ID）。
写入时按需 allocate，Attention 时按页表寻址组合计算，释放时回收物理块。

运行
----
    cd LLM
    python3 paged_attention.py
"""

from __future__ import annotations

from typing import List

import torch
import torch.nn as nn
import torch.nn.functional as F


# ==========================================
# 1. 全局物理显存池与中央管理器 (Block Manager)
# ==========================================
class BlockManager:
    def __init__(self, num_blocks: int):
        self.num_blocks = num_blocks
        self.free_blocks: List[int] = list(range(num_blocks))
        self.ref_counts = {block_id: 0 for block_id in range(num_blocks)}

    def allocate_block(self) -> int:
        """物理显存分配：从空闲列表中弹出一个不连续的物理 Block ID。"""
        if not self.free_blocks:
            raise MemoryError("GPU HBM Out of Memory! 全局物理显存池已耗尽。")
        block_id = self.free_blocks.pop(0)
        self.ref_counts[block_id] = 1
        return block_id

    def free_block(self, block_id: int) -> None:
        """物理显存回收：将不再使用的物理 Block ID 放回空闲列表。"""
        if self.ref_counts[block_id] > 0:
            self.ref_counts[block_id] -= 1
            if self.ref_counts[block_id] == 0:
                self.free_blocks.append(block_id)
                self.free_blocks.sort()

    def get_num_free_blocks(self) -> int:
        return len(self.free_blocks)


# ==========================================
# 2. 具备动态分配功能的推理引擎
# ==========================================
class AdvancedPagedAttentionEngine(nn.Module):
    def __init__(self, block_manager: BlockManager, block_size: int,
                 num_heads: int, head_dim: int):
        super().__init__()
        self.block_manager = block_manager
        self.block_size = block_size
        self.num_heads = num_heads
        self.head_dim = head_dim

        # 静态分配全局 GPU HBM 显存池
        # [总Block数, 2(K/V), num_heads, block_size, head_dim]
        self.gpu_kv_buffer = torch.zeros(
            (block_manager.num_blocks, 2, num_heads, block_size, head_dim)
        )

    def append_and_manage_kv(
        self,
        k: torch.Tensor,
        v: torch.Tensor,
        block_table: List[int],
        global_seq_len: int,
        verbose: bool = True,
    ) -> None:
        """
        核心动态写入：检查是否需要新物理块，并离散写入 K/V。
        k/v: [batch, num_heads, 1, head_dim]
        """
        if global_seq_len % self.block_size == 0:
            new_physical_id = self.block_manager.allocate_block()
            block_table.append(new_physical_id)
            if verbose:
                print(
                    f"  [显存动态分配] 触发新块申请！指派物理 Block ID: {new_physical_id}. "
                    f"当前页表更新为: {block_table}"
                )

        logical_block_idx = global_seq_len // self.block_size
        block_offset = global_seq_len % self.block_size
        physical_block_id = block_table[logical_block_idx]

        # 原地离散写入（Zero Copy）
        self.gpu_kv_buffer[physical_block_id, 0, :, block_offset, :] = k[0, :, 0, :]
        self.gpu_kv_buffer[physical_block_id, 1, :, block_offset, :] = v[0, :, 0, :]

    def paged_attention_forward(
        self,
        q: torch.Tensor,
        block_table: List[int],
        current_seq_len: int,
    ) -> torch.Tensor:
        """根据离散 block_table 执行组合 Attention 计算。"""
        _, num_heads, _, head_dim = q.shape
        q_s = q.squeeze(0).squeeze(1)  # [num_heads, head_dim]

        global_context = torch.zeros((num_heads, 1, head_dim), device=q.device, dtype=q.dtype)
        all_token_logits: List[List[torch.Tensor]] = [[] for _ in range(num_heads)]

        # 步骤 A: 物理不连续寻址，计算 Logits
        token_counter = 0
        for physical_block_id in block_table:
            rem_tokens = current_seq_len - token_counter
            actual_tokens_in_block = min(self.block_size, rem_tokens)
            if actual_tokens_in_block <= 0:
                break

            block_k = self.gpu_kv_buffer[
                physical_block_id, 0, :, :actual_tokens_in_block, :
            ]

            for h in range(num_heads):
                block_logits = torch.matmul(
                    q_s[h].unsqueeze(0), block_k[h].transpose(-2, -1)
                ) / (head_dim ** 0.5)
                all_token_logits[h].append(block_logits.squeeze(0))
            token_counter += actual_tokens_in_block

        # 步骤 B: 跨离散块的全局 Softmax 与 Value 加权聚合
        token_counter = 0
        for physical_block_id in block_table:
            rem_tokens = current_seq_len - token_counter
            actual_tokens_in_block = min(self.block_size, rem_tokens)
            if actual_tokens_in_block <= 0:
                break

            block_v = self.gpu_kv_buffer[
                physical_block_id, 1, :, :actual_tokens_in_block, :
            ]

            for h in range(num_heads):
                full_logits_this_head = torch.cat(all_token_logits[h])
                attn_weights_this_head = F.softmax(full_logits_this_head, dim=-1)
                current_block_weights = attn_weights_this_head[
                    token_counter : token_counter + actual_tokens_in_block
                ]
                global_context[h] += torch.matmul(
                    current_block_weights.unsqueeze(0), block_v[h]
                )
            token_counter += actual_tokens_in_block

        return global_context.unsqueeze(0)  # [1, num_heads, 1, head_dim]


def demo() -> None:
    torch.manual_seed(0)

    # 全局共有 5 个物理 Blocks，单块存 2 个 Token；2 个头，维度 64
    global_block_manager = BlockManager(num_blocks=5)
    engine = AdvancedPagedAttentionEngine(
        global_block_manager, block_size=2, num_heads=2, head_dim=64
    )

    print(
        f"系统初始空闲物理块数量: {global_block_manager.get_num_free_blocks()} 个, "
        f"列表: {global_block_manager.free_blocks}"
    )

    # 场景一：用户 A 自回归生成 3 个 Token
    print("\n=== 场景一：用户 A 请求进入 ===")
    user_a_block_table: List[int] = []
    user_a_seq_len = 0

    for step in range(3):
        print(f"用户 A 生成第 {step + 1} 个 Token:")
        q = torch.randn(1, 2, 1, 64)
        k = torch.randn(1, 2, 1, 64)
        v = torch.randn(1, 2, 1, 64)

        engine.append_and_manage_kv(k, v, user_a_block_table, global_seq_len=user_a_seq_len)
        user_a_seq_len += 1
        out = engine.paged_attention_forward(q, user_a_block_table, current_seq_len=user_a_seq_len)
        assert out.shape == (1, 2, 1, 64)

    print(f"用户 A 结束。最终占用物理页表: {user_a_block_table}")
    print(
        f"此时系统剩余空闲物理块数量: {global_block_manager.get_num_free_blocks()} 个, "
        f"列表: {global_block_manager.free_blocks}"
    )
    assert len(user_a_block_table) == 2
    assert global_block_manager.get_num_free_blocks() == 3

    # 场景二：用户 B 并发进入，生成 2 个 Token
    print("\n=== 场景二：用户 B 并发请求进入 ===")
    user_b_block_table: List[int] = []
    user_b_seq_len = 0

    for step in range(2):
        print(f"用户 B 生成第 {step + 1} 个 Token:")
        q = torch.randn(1, 2, 1, 64)
        k = torch.randn(1, 2, 1, 64)
        v = torch.randn(1, 2, 1, 64)
        engine.append_and_manage_kv(k, v, user_b_block_table, global_seq_len=user_b_seq_len)
        user_b_seq_len += 1
        engine.paged_attention_forward(q, user_b_block_table, current_seq_len=user_b_seq_len)

    print(f"用户 B 结束。最终占用物理页表: {user_b_block_table}")
    print(f"此时系统剩余空闲物理块数量: {global_block_manager.get_num_free_blocks()} 个")
    assert len(user_b_block_table) == 1
    assert global_block_manager.get_num_free_blocks() == 2

    # 场景三：用户 A 结束，显存回收
    print("\n=== 场景三：用户 A 彻底结束，触发显存动态回收 ===")
    print(f"回收前系统空闲列表: {global_block_manager.free_blocks}")

    for block_id in user_a_block_table:
        global_block_manager.free_block(block_id)
    user_a_block_table = []

    print(f"回收后系统空闲列表: {global_block_manager.free_blocks}")
    print(
        f"此时系统剩余空闲物理块数量: {global_block_manager.get_num_free_blocks()} 个"
        f"（用户 A 释放的块已能无缝被新请求复用）"
    )
    assert global_block_manager.get_num_free_blocks() == 4
    print("\nPagedAttention demo passed.")


if __name__ == "__main__":
    demo()

File: LLM/paged_attention.cpp
// PagedAttention — discrete physical KV blocks with per-request block tables.
//
// Mirrors LLM/paged_attention.py using plain C++ (no PyTorch).
//
// Whiteboard talking points:
// - Logical KV sequence is split into fixed-size blocks; block_table maps
//   logical block index -> physical block id in a global HBM pool.
// - append_and_manage_kv allocates non-contiguous blocks on demand (no torch.cat).
// - paged_attention_forward gathers K/V via the page table, then global softmax.
// - free_block returns physical ids to the pool for reuse (vLLM-style serving).

#include <algorithm>
#include <cassert>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <random>
#include <stdexcept>
#include <vector>

class BlockManager {
 public:
    explicit BlockManager(int num_blocks) : ref_counts_(num_blocks, 0) {
        free_blocks_.reserve(num_blocks);
        for (int i = 0; i < num_blocks; ++i) {
            free_blocks_.push_back(i);
        }
    }

    int allocate_block() {
        if (free_blocks_.empty()) {
            throw std::runtime_error("GPU HBM Out of Memory! global block pool exhausted");
        }
        const int block_id = free_blocks_.front();
        free_blocks_.erase(free_blocks_.begin());
        ref_counts_[static_cast<size_t>(block_id)] = 1;
        return block_id;
    }

    void free_block(int block_id) {
        if (ref_counts_[static_cast<size_t>(block_id)] > 0) {
            --ref_counts_[static_cast<size_t>(block_id)];
            if (ref_counts_[static_cast<size_t>(block_id)] == 0) {
                free_blocks_.push_back(block_id);
                std::sort(free_blocks_.begin(), free_blocks_.end());
            }
        }
    }

    int num_free_blocks() const { return static_cast<int>(free_blocks_.size()); }

    const std::vector<int>& free_blocks() const { return free_blocks_; }

 private:
    std::vector<int> free_blocks_;
    std::vector<int> ref_counts_;
};

class GpuKvBuffer {
 public:
    GpuKvBuffer(int num_blocks, int block_size, int num_heads, int head_dim)
        : block_size_(block_size),
          num_heads_(num_heads),
          head_dim_(head_dim),
          data_(static_cast<size_t>(num_blocks) * 2 * num_heads * block_size * head_dim,
                0.0f) {}

    float& at(int block, int kv, int head, int offset, int dim) {
        return data_[index(block, kv, head, offset, dim)];
    }

    float at(int block, int kv, int head, int offset, int dim) const {
        return data_[index(block, kv, head, offset, dim)];
    }

    int block_size() const { return block_size_; }
    int num_heads() const { return num_heads_; }
    int head_dim() const { return head_dim_; }

 private:
    int block_size_;
    int num_heads_;
    int head_dim_;
    std::vector<float> data_;

    size_t index(int block, int kv, int head, int offset, int dim) const {
        return (((static_cast<size_t>(block) * 2 + kv) * num_heads_ + head) * block_size_ +
                offset) *
                   head_dim_ +
               dim;
    }
};

class AdvancedPagedAttentionEngine {
 public:
    AdvancedPagedAttentionEngine(BlockManager& manager, int num_blocks, int block_size,
                                 int num_heads, int head_dim)
        : manager_(manager),
          block_size_(block_size),
          num_heads_(num_heads),
          head_dim_(head_dim),
          buffer_(num_blocks, block_size, num_heads, head_dim) {}

    void append_and_manage_kv(const std::vector<float>& k, const std::vector<float>& v,
                              std::vector<int>& block_table, int global_seq_len,
                              bool verbose = true) {
        if (global_seq_len % block_size_ == 0) {
            const int new_id = manager_.allocate_block();
            block_table.push_back(new_id);
            if (verbose) {
                std::cout << "  [alloc] new physical block id: " << new_id << ", table: [";
                for (size_t i = 0; i < block_table.size(); ++i) {
                    if (i > 0) {
                        std::cout << ", ";
                    }
                    std::cout << block_table[i];
                }
                std::cout << "]\n";
            }
        }

        const int logical_idx = global_seq_len / block_size_;
        const int offset = global_seq_len % block_size_;
        const int physical_id = block_table[static_cast<size_t>(logical_idx)];

        for (int h = 0; h < num_heads_; ++h) {
            for (int d = 0; d < head_dim_; ++d) {
                buffer_.at(physical_id, 0, h, offset, d) =
                    k[static_cast<size_t>(h) * head_dim_ + d];
                buffer_.at(physical_id, 1, h, offset, d) =
                    v[static_cast<size_t>(h) * head_dim_ + d];
            }
        }
    }

    std::vector<float> paged_attention_forward(const std::vector<float>& q,
                                               const std::vector<int>& block_table,
                                               int current_seq_len) const {
        const float scale = 1.0f / std::sqrt(static_cast<float>(head_dim_));
        std::vector<std::vector<float>> all_logits(static_cast<size_t>(num_heads_));

        int token_counter = 0;
        for (int physical_id : block_table) {
            const int rem = current_seq_len - token_counter;
            const int actual = std::min(block_size_, rem);
            if (actual <= 0) {
                break;
            }

            for (int h = 0; h < num_heads_; ++h) {
                for (int t = 0; t < actual; ++t) {
                    float dot = 0.0f;
                    for (int d = 0; d < head_dim_; ++d) {
                        dot += q[static_cast<size_t>(h) * head_dim_ + d] *
                               buffer_.at(physical_id, 0, h, t, d);
                    }
                    all_logits[static_cast<size_t>(h)].push_back(dot * scale);
                }
            }
            token_counter += actual;
        }

        std::vector<std::vector<float>> attn_weights(static_cast<size_t>(num_heads_));
        for (int h = 0; h < num_heads_; ++h) {
            const auto& logits = all_logits[static_cast<size_t>(h)];
            float max_v = logits.empty() ? 0.0f : logits[0];
            for (float v : logits) {
                max_v = std::max(max_v, v);
            }
            attn_weights[static_cast<size_t>(h)].resize(logits.size());
            float sum = 0.0f;
            for (size_t i = 0; i < logits.size(); ++i) {
                attn_weights[static_cast<size_t>(h)][i] = std::exp(logits[i] - max_v);
                sum += attn_weights[static_cast<size_t>(h)][i];
            }
            for (float& w : attn_weights[static_cast<size_t>(h)]) {
                w /= sum;
            }
        }

        std::vector<float> context(static_cast<size_t>(num_heads_) * head_dim_, 0.0f);
        token_counter = 0;
        for (int physical_id : block_table) {
            const int rem = current_seq_len - token_counter;
            const int actual = std::min(block_size_, rem);
            if (actual <= 0) {
                break;
            }

            for (int h = 0; h < num_heads_; ++h) {
                for (int t = 0; t < actual; ++t) {
                    const float w =
                        attn_weights[static_cast<size_t>(h)][static_cast<size_t>(token_counter + t)];
                    for (int d = 0; d < head_dim_; ++d) {
                        context[static_cast<size_t>(h) * head_dim_ + d] +=
                            w * buffer_.at(physical_id, 1, h, t, d);
                    }
                }
            }
            token_counter += actual;
        }

        return context;
    }

 private:
    BlockManager& manager_;
    int block_size_;
    int num_heads_;
    int head_dim_;
    GpuKvBuffer buffer_;
};

static std::vector<float> random_kv(int num_heads, int head_dim, std::mt19937& rng) {
    std::normal_distribution<float> dist(0.0f, 1.0f);
    std::vector<float> x(static_cast<size_t>(num_heads) * head_dim);
    for (float& v : x) {
        v = dist(rng);
    }
    return x;
}

static void print_free_blocks(const BlockManager& manager) {
    std::cout << "free blocks (" << manager.num_free_blocks() << "): [";
    const auto& blocks = manager.free_blocks();
    for (size_t i = 0; i < blocks.size(); ++i) {
        if (i > 0) {
            std::cout << ", ";
        }
        std::cout << blocks[i];
    }
    std::cout << "]\n";
}

static void demo() {
    constexpr int kNumBlocks = 5;
    constexpr int kBlockSize = 2;
    constexpr int kNumHeads = 2;
    constexpr int kHeadDim = 64;

    BlockManager manager(kNumBlocks);
    AdvancedPagedAttentionEngine engine(manager, kNumBlocks, kBlockSize, kNumHeads, kHeadDim);
    std::mt19937 rng(0);

    std::cout << "Initial ";
    print_free_blocks(manager);

    std::cout << "\n=== User A ===\n";
    std::vector<int> user_a_table;
    int user_a_len = 0;
    for (int step = 0; step < 3; ++step) {
        std::cout << "User A token " << (step + 1) << ":\n";
        const std::vector<float> q = random_kv(kNumHeads, kHeadDim, rng);
        const std::vector<float> k = random_kv(kNumHeads, kHeadDim, rng);
        const std::vector<float> v = random_kv(kNumHeads, kHeadDim, rng);
        engine.append_and_manage_kv(k, v, user_a_table, user_a_len);
        ++user_a_len;
        const std::vector<float> out =
            engine.paged_attention_forward(q, user_a_table, user_a_len);
        assert(static_cast<int>(out.size()) == kNumHeads * kHeadDim);
    }
    std::cout << "User A table: [";
    for (size_t i = 0; i < user_a_table.size(); ++i) {
        if (i > 0) {
            std::cout << ", ";
        }
        std::cout << user_a_table[i];
    }
    std::cout << "], free=" << manager.num_free_blocks() << "\n";
    assert(user_a_table.size() == 2);
    assert(manager.num_free_blocks() == 3);

    std::cout << "\n=== User B ===\n";
    std::vector<int> user_b_table;
    int user_b_len = 0;
    for (int step = 0; step < 2; ++step) {
        std::cout << "User B token " << (step + 1) << ":\n";
        const std::vector<float> q = random_kv(kNumHeads, kHeadDim, rng);
        const std::vector<float> k = random_kv(kNumHeads, kHeadDim, rng);
        const std::vector<float> v = random_kv(kNumHeads, kHeadDim, rng);
        engine.append_and_manage_kv(k, v, user_b_table, user_b_len);
        ++user_b_len;
        engine.paged_attention_forward(q, user_b_table, user_b_len);
    }
    std::cout << "User B table size=" << user_b_table.size()
              << ", free=" << manager.num_free_blocks() << "\n";
    assert(user_b_table.size() == 1);
    assert(manager.num_free_blocks() == 2);

    std::cout << "\n=== Release User A ===\n";
    print_free_blocks(manager);
    for (int block_id : user_a_table) {
        manager.free_block(block_id);
    }
    user_a_table.clear();
    print_free_blocks(manager);
    assert(manager.num_free_blocks() == 4);
}

int main() {
    demo();
    std::cout << "paged_attention: all tests passed\n";
    return 0;
}
